Fix crash when arguments of the dictionary layout are zero values

ahmadov opened the pull request:
Fixes #78314 

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Dictionaries of type `ssd_cache` now reject zero or negative `block_size` and `write_buffer_size` parameters (issue #78314).

ahmadov committed:
Fix crash when arguments of the dictionary layout are zero values

The `block_size` and `write_buffer_size` are arguments to the
`ssd_cache` layout of the dictionary. They must be non-zero values.

Fixes #78314

rschu1ze commented on src/Dictionaries/getDictionaryConfigurationFromAST.cpp:
```diff
@@ -166,6 +166,13 @@ void buildLayoutConfiguration(
                 value_field.getTypeName());
         }
 
+        if (value_field.safeGet<UInt64>() == 0)
```
Did you test what happens if a negative value is passed as Float or String (--> l. 161), e.g. `-2.0`? I could imagine that `safeGet<UInt64>` does a numeric roll-over internally and we'll get UINT_MAX - 1 in which case the new check would not work.

Please also add a small SQL test --> https://clickhouse.com/docs/development/tests in tests/queries/0_stateless/ - thanks!

ahmadov replied:
yes, I tested with negative value and it fails in the previous check. It detects the negative value as `Int64`, then it throws an error.

ahmadov replied:
added the test with cc40fc3c996d4eb09f93776d6ef02c435cc78b93

ahmadov committed:
add test

ahmadov committed:
add one more test

ahmadov committed:
Add check for dictionary lifetime min and max values

Throws an error in case `MIN` value is greater than the `MAX` value.

ahmadov committed:
fix tests

Also, limit the zero value check only to the `ssd_cache` layout.

rschu1ze commented on tests/queries/0_stateless/03405_ssd_cache_layout.sh:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+# Tags: no-replicated-database
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
```
@ahmadov .sql tests are simpler than .sh tests (fewer degrees of freedom, less complexity), and therefore preferred. In some cases .sh tests are needed but as far as I see, this isn't one of them.

It is possible to do "negative" tests in .sql tests as well - see tests/queries/0_stateless/02354_vector_search_index_creation_negative.sql for an example. One can only check the error code instead the full error string but that is fine.

Therefore, could you please try to convert this test to .sql? Thanks!

ahmadov replied:
yes, the only reason I used `.sh` tests to check the exact error message but as you mentioned it's not so important since the error codes can be used as well. Moved them into `.sql` tests with 1d06a0c061cc1822e6e1a57db4e36a3f938428ad

rschu1ze commented on tests/queries/0_stateless/03405_ssd_cache_layout.sh:
```diff
@@ -0,0 +1,44 @@
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
```
Let's also add a comment on top which states that this is a test for GitHub issue #78314

ahmadov replied:
added in 1d06a0c061cc1822e6e1a57db4e36a3f938428ad

rschu1ze commented on tests/queries/0_stateless/03406_dictionary_lifetime.sh:
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Same comments as in the other test apply here.

Minor comment: Could we maybe rename the test to 03406_dictionary_incorrect_min_max_lifetimes?

rschu1ze replied:
Let's rename the other test similarly: tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sh

ahmadov committed:
address pr comments

ahmadov commented on src/Dictionaries/getDictionaryConfigurationFromAST.cpp:
```diff
@@ -81,6 +81,14 @@ void buildLifetimeConfiguration(
+    {
+        throw DB::Exception(
+            ErrorCodes::BAD_ARGUMENTS,
+            "{} parameter 'MIN' must be less than or equal to 'MAX'.",
```
1d06a0c061cc1822e6e1a57db4e36a3f938428ad

ahmadov committed:
fix build

rschu1ze commented on tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sql:
```diff
@@ -0,0 +1,18 @@
+DROP DICTIONARY IF EXISTS dx;
+
+SELECT 'BLOCK_SIZE is a negative value.';
+CREATE DICTIONARY dx (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(BLOCK_SIZE -1)) LIFETIME(1); -- { serverError 36 }
```
`36` will work but the cleartext error name from src/Common/ErrorCodes.cpp would also work:

```cpp
-- { serverError BAD_ARGUMENTS }
```

ahmadov replied:
a833c89a8e6cae4b1a2066122217f036d9361f30

rschu1ze replied:
Cool, thanks. No need to post commit SHAs (I wouldn't usually check the individual commits anyways - I just look at the entire PR) 😄 Rather, feel free to mark the comment as resolved.

ahmadov replied:
kinda old habits 😄

ahmadov committed:
prefer error names instead of error codes for better readability

ahmadov committed:
fix ambiguous `UInt64` type

The same type is also defined as `Poco::UInt64`.

Prefer to use the phrase "value should be positive number" since it's
already used in other error messages.
